client/solve.go

```diff
@@ -255,19 +282,49 @@ func (c *Client) solve(ctx context.Context, def *llb.Definition, runGateway runG
 			frontendInputs[key] = def.ToPB()
 		}
 
+		exports := make([]*controlapi.Exporter, 0, len(opt.Exports))
+		var localExporter *controlapi.Exporter
+		for _, exp := range exporters {
+			if exp.Type != ExporterLocal {
+				exports = append(exports, &controlapi.Exporter{
+					ID:    exp.id,
+					Type:  exp.Type,
+					Attrs: exp.Attrs,
+				})
+			} else if localExporter == nil {
+				// TODO(dima): different options per local exporter?
+				localExporter = &controlapi.Exporter{
+					Type:  ExporterLocal,
+					Attrs: exp.Attrs,
+				}
+			}
+		}
+		if localExporter != nil {
+			// Add a single instance of the local exporter
+			// since it's replicated entirely on the client
+			exports = append(exports, localExporter)
+		}
+
+		// For older server, use the first exporter
+		var ex ExportEntry
+		if len(opt.Exports) != 0 {
+			ex = opt.Exports[0]
+		}
+
 		resp, err := c.ControlClient().Solve(ctx, &controlapi.SolveRequest{
-			Ref:            ref,
-			Definition:     pbd,
-			Exporter:       ex.Type,
-			ExporterAttrs:  ex.Attrs,
-			Session:        s.ID(),
-			Frontend:       opt.Frontend,
-			FrontendAttrs:  frontendAttrs,
-			FrontendInputs: frontendInputs,
-			Cache:          cacheOpt.options,
-			Entitlements:   opt.AllowedEntitlements,
-			Internal:       opt.Internal,
-			SourcePolicy:   opt.SourcePolicy,
+			Ref:                     ref,
+			Definition:              pbd,
+			ExporterDeprecated:      ex.Type,
+			ExporterAttrsDeprecated: ex.Attrs,
```

Comment:
If this is deprecated, can we just reduce these fields to `Exporters` and remove them from `SolveRequest` like you have done for `ExporterRequest`? It will be easier to handle in the control api I think.

Reply:
I thought this to support the new client old server type of scenario. I agree that it's an unlikely one but still possible.
I can remove them though.